homeassistant/components/mobile_app/push_notification.py

```diff
@@ -0,0 +1,90 @@
+"""Push notification handling."""
+import asyncio
+from typing import Callable
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.event import async_call_later
+from homeassistant.util.uuid import random_uuid_hex
+
+PUSH_CONFIRM_TIMEOUT = 10  # seconds
+
+
+class PushChannel:
+    """Class that represents a push channel."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        webhook_id: str,
+        support_confirm: bool,
+        send_message: Callable[[dict], None],
+        on_teardown: Callable[[], None],
+    ) -> None:
+        """Initialize a local push channel."""
+        self.hass = hass
+        self.webhook_id = webhook_id
+        self.support_confirm = support_confirm
+        self._send_message = send_message
+        self.on_teardown = on_teardown
+        self.pending_confirms = {}
+
+    @callback
+    def async_send_notification(self, data, fallback_send):
+        """Send a push notification."""
+        if not self.support_confirm:
+            self._send_message(data)
+            return
+
+        confirm_id = random_uuid_hex()
+        data["hass_confirm_id"] = confirm_id
+
+        async def handle_push_failed(_=None):
+            """Handle a failed local push notification."""
+            # Remove this handler from the pending dict
+            # If it didn't exist we hit a race condition between call_later and another
+            # push failing and tearing down the connection.
```

Comment:
Can the problem be clarified a bit?

This race is clear: `handle_push_failed` is called after `async_confirm_notification`, that's handled by checking if `confirm_id` is no longer in the `self.pending_confirms` dict

What's this race though:
> race condition between call_later and another push failing and tearing down the connection

Is that if `handle_push_failed` is called by two different failing pushes?

Reply:
there is a very rare condition possible where call_later of a failed connection is scheduled, but the connection is already started tearing down before it was able to undo the `call_later` call. The tearing down can be caused by a) another message not being confirmed or b) websocket connection being dropped.